### DISCUSSION TITLE
Is it possible to write unit tests for TCA?

### DISCUSSION THREAD

**Discussion by theMuslimDev** [pointfreeco/swift-composable-architecture#1975] [Q&A] [answered]
Just out of curiosity is it possible to write unit tests using TCA?

**Comment by Alex293** [comment]
Testing is one of the 5 goals according to the first paragraph of the read me.
I should add that Unit testing in TCA is a joy with so many powerful tools shipped with the libs.

**Comment by mbrandonw** [comment]
Hi @Muhammed9991, as @Alex293 said, testing is a huge part of TCA. [Here's the documentation](https://pointfreeco.github.io/swift-composable-architecture/main/documentation/composablearchitecture/testing) on how to test with the library.

**Comment by theMuslimDev** [comment]
Definitely agree with you @Alex293. I enjoy writing tests with TCA. However, they are more of integration tests as opposed to unit tests. I was just wondering if its possible to write unit tests with TCA @mbrandonw.

**Reply by mbrandonw** [reply]
Ah, well perhaps we differ on the definition of unit test, but in my mind all TCA tests are unit tests. Sure you can test multiple features at once, and hence in some sense the "integration" of the features, but that is not what most people think of when they say "integration test." I believe that phrase is mostly used when testing code that interacts with outside systems, e.g. making live network requests, running code in a host (simulator, browser, etc.), and so on.

A unit test is usually meant to convey a test that runs in complete isolation and only involves the types and functions that you control. So, it doesn't run in a host, and doesn't involve outside systems or dependencies.

Just as a silly example, consider writing a test that calls out to another function:

```swift
func foo() -> Int {
  bar() + 1
}
func bar() -> Int { … }
```

If you wrote a test for `foo` you would also be exercising the logic of `bar`, but you wouldn't call that an "integration" test. We feel that same applies to writing tests for multiple composed reducers. After all, reducers are just functions and composing them just means one reducer calls out to another. So testing it does exercise their integration, but not in the sense that "integration test" is normally thrown around.

**Reply by theMuslimDev** [reply]
Ohh. You're right. I got confused from this part in the README for TCA.


<img width="789" alt="Screenshot 2023-03-14 at 15 30 37" src="https://user-images.githubusercontent.com/80204376/225052494-0c7594c9-5138-453b-8c59-495845c7d8b9.png">

But it makes sense that they are essentially unit tests.

**Reply by mbrandonw** [reply]
Haha yeah I can see how it may be confusing, but essentially what we are trying to say is that TCA allows you to easily write tests that involve many parts of an application (multiple features + side effects), but all in a unit testing style.

**Comment by theMuslimDev** [comment, marked as answer]
Reply by @mbrandonw answers this. I can't mark it as the correct answer because it's in a thread.

> Ah, well perhaps we differ on the definition of unit test, but in my mind all TCA tests are unit tests. Sure you can test multiple features at once, and hence in some sense the "integration" of the features, but that is not what most people think of when they say "integration test." I believe that phrase is mostly used when testing code that interacts with outside systems, e.g. making live network requests, running code in a host (simulator, browser, etc.), and so on.
> 
> 
> 
> A unit test is usually meant to convey a test that runs in complete isolation and only involves the types and functions that you control. So, it doesn't run in a host, and doesn't involve outside systems or dependencies.
> 
> 
> 
> Just as a silly example, consider writing a test that calls out to another function:
> 
> 
> 
> ```swift
> 
> func foo() -> Int {
> 
>   bar() + 1
> 
> }
> 
> func bar() -> Int { … }
> 
> ```
> 
> 
> 
> If you wrote a test for `foo` you would also be exercising the logic of `bar`, but you wouldn't call that an "integration" test. We feel that same applies to writing tests for multiple composed reducers. After all, reducers are just functions and composing them just means one reducer calls out to another. So testing it does exercise their integration, but not in the sense that "integration test" is normally thrown around.
